### DISCUSSION TITLE
How can I access the context from the component function?

### DISCUSSION THREAD

**Discussion by sawa-ko** [TanStack/router#1628] [Q&A] [answered]
I am currently defining a default context when creating the router, but I have no way to access it through the component. What I have to do is use the loader function, and then use the hook useLoaderData inside the component. It is very cumbersome to do and repetitive and I would like that through the arguments passed to the component function I get the context instead of doing all those steps above each time and so I also avoid using unnecessary hooks.

```tsx
const router = createRouter({
  // eslint-disable-next-line react/prop-types
  Wrap: ({ children }) => {
    return <ConfigProvider theme={{ cssVar: true }}>{children}</ConfigProvider>;
  },
  context: {
    theme: theme.getDesignToken({ cssVar: true }),
  },
  routeTree,
});
```

```tsx
export const Route = createRootRouteWithContext<{ theme: AliasToken }>()({
  component: Root,
});

export function Root() {
  return (
    <>
      <Outlet />
      <TanStackRouterDevtools />
    </>
  );
}
```

```tsx
// Current solution
export const Route = createFileRoute('/')({
  component: () => {
    return (
      <>
        <HomePage />
      </>
    );
  },
  loader: ({ context }) => {
    return {
      ...context,
    };
  },
});

export function HomePage() {
  const x = Route.useLoaderData();

  return (
    <div>
      <h3>Welcome Home! {x.theme.blue}</h3>
    </div>
  );
}

// Better solution
export const Route = createFileRoute('/')({
  component: () => {
    return (
      <>
        <HomePage />
      </>
    );
  }
});

export function HomePage({ context: x }) {

  return (
    <div>
      <h3>Welcome Home! {x.theme.blue}</h3>
    </div>
  );
}
```

**Comment by TkDodo** [comment]
It's available on matches, so useRouterState or useMatches should give you the context for each match:

https://tanstack.com/router/latest/docs/framework/react/guide/router-context#processing-accumulated-route-context

**Reply by mindaugasb-isense** [reply]
But it's not exactly the same.

**Reply by schiller-manuel** [reply]
not the same as what?

**Comment by sawa-ko** [comment]
It's not working:

![image](https://github.com/TanStack/router/assets/56084970/b48bc703-ecae-4bff-96d3-1a71fa101674)

**Comment by sawa-ko** [comment]
@TkDodo

**Comment by sawa-ko** [comment]
I see that with useMatches it comes in the loaderData property, but as I mentioned, isn't there a way to do it without hooks?

Let the component that renders the path receive that as props.

**Reply by TkDodo** [reply, marked as answer]
components that render on a route don't receive any props. You get loader data with `useLoaderData()`. You get path params with `useParams()`. And you can get context with `useMatches()` and then access `.context`. It's even shown in your image ...

**Reply by dylanhu7** [reply]
@TkDodo Just curious - what is the reason that the `component` of the route doesn't receive `loaderData` (and potentially other stuff like `context`) as props? Is it just to have a unified API through the more flexible hooks instead of having both DI through props and through hooks?

**Reply by TkDodo** [reply]
if you inline the component, sure, this sounds nice, but once you extract it you need to type the props and lose type inference. And you need the hooks anyways, otherwise you’re stuck with prop drilling from the component downwards.

**Reply by dylanhu7** [reply]
Makes sense, thanks. If you extract the `component`, would it make sense to type the props with something like
```tsx
function MyComponent({ loaderData, context }: typeof Route.ComponentProps)
```
(looking very much like RRv7 haha)

Not pushing for this, just curious about the considerations in choosing to implement or not implement this.

**Reply by thomastvedt** [reply]
```tsx
  const accumulatedContext = useMatches({
    structuralSharing: true,
    select: (matches) => matches[matches.length - 1].context
  });
```
So this is the way to get the router context in a component? I was hoping there would be a `useRouterContext()` hook that I could easily plugin in my components :) 

My scenario is something like this:
- there is a tree structure root->retailer->machine
- url /machine/123 must resolve the machine's retailerId (beforeLoad, async query, setting retailerId on context)
- I don't use the built in data loader functions, but simply load data in the component. 
- in my breadcrumbs I need both the machineId (param) and the resolved retailerId (context)

Questions:
- Is the router context meant mostly for the data loader functions?
- Would you recommend to roll my own context separate from the router context for my scenario?

**Reply by TkDodo** [reply]
> I was hoping there would be a useRouterContext() hook

You want `useRouteContext({ from: '/my-route/' })`

**Reply by thomastvedt** [reply]
> > I was hoping there would be a useRouterContext() hook
> 
> You want `useRouteContext({ from: '/my-route/' })`

If I understand correctly that will only return the context accumulated until the specified route? I' looking for the "complete accumulated context" for the current resolved route?

In my App route (MainLayout) I'm interested in the complete router context. If I'm at url /_app/retailer/123, I want to render some info based on the context in the _app layout.

Some test code:

```tsx
export const Route = createFileRoute('/_app')({
  component: MainLayout,
})

function MainLayout() {
  const accumulatedContext = useMatches({
    structuralSharing: true,
    select: (matches) => matches[matches.length - 1].context
  });

  const contextFromAppRoute = Route.useRouteContext();
  const anotherContextTest = useRouteContext({
    from: '/_app'
  });
  
  const info = {
    category,
    level,
    accumulatedContext,
    contextButNotAccumulated,
    anotherContextTest
  };
```

result:

<img width="289" alt="image" src="https://github.com/user-attachments/assets/7852ce67-9cbb-4e96-b104-3d10fb492b2f" />

**Reply by TkDodo** [reply]
oh right, yeah I think you need `useMatches` for that (or `useRouterState`)

**Reply by jedrichards** [reply]
@TkDodo 

> if you inline the component, sure, this sounds nice, but once you extract it you need to type the props and lose type inference. And you need the hooks anyways, otherwise you’re stuck with prop drilling from the component downwards.

What about a pattern like:

```ts
export const Route = createFileRoute('/blah/$id')({
  component: ({ params }) => {
    return <MyComponent id={params.id} />
  }
})
```

this seems more typesafe to me, because you are using the inferred params directly, within the route config. The alternative, `Route.useParams()` seems a bit more hand-wavy, since you're not tying the Route and it's component implementation together fundamentally, you're relying on importing and using the correct `Route`, and maintaining that fuzzy connection over time. You could end up with a situation of a component importing effectively an unrelated route, but because of mere temporary structural similarlity between the route params, and the component requirements, you end up with no compile errors.

The pattern above, keeps the inner component `MyComponent` a bit less coupled to routing in general. It just expresses its props requriements in router agnostic terms, but its props API is not less typesafe because of it.

`Route.useParams()` is a nice idea, and I'm sure not everybody will like to write little prop forwarding wrappers in the route config, but is there a downside for making this kind of API available?

**Reply by TkDodo** [reply]
you don’t have to use it outside of your route if you don’t want to. The pattern you’ve described works fine for one layer of prop drilling, but we all know it won’t stay that way. I’ve personally witnessed apps where we passed those path params down 7+ layers. It’s a mess to get to understand “where does this id come from?”. Reading it form `Route.useParams` ties it to a route, sure, but this can be seen as an advantage because you immediately know where things come from when reading a component in isolation.

As for your pattern, you can do:

```
export const Route = createFileRoute('/blah/$id')({
  component: RouteComponent
})

const RouteComponent = () => {
  return <MyComponent id={Route.useParams().id} />
}
```
